### PR TITLE
Fix storing more than ~15000 levels of experience

### PR DESCRIPTION
**Pull request by minecraft552**
When you store too much experience, it would overflow into negative numbers. I changed the necessary numbers to be `long` type instead of `int` in order to prevent this.

The reasoning is that in vanilla survival it is (technically) possible to get way more than the ~15000 that is possible with the `int` type.

### CODE REVIEW

File: src/main/java/bl4ckscor3/mod/globalxp/openmods/utils/EnchantmentUtils.java [hunk 1]
```diff
@@ -23,15 +23,15 @@ private EnchantmentUtils() {}
...
 
-	public static void addPlayerXP(Player player, int amount) {
-		int experience = getPlayerXP(player) + amount;
-		int expForLevel;
+	public static void addPlayerXP(Player player, long amount) {
+		long experience = getPlayerXP(player) + amount;
+		long expForLevel;
 
-		player.totalExperience = experience;
+		player.totalExperience = (int)Math.min(Integer.MAX_VALUE, experience);
```
Comment (bl4ckscor3) [inline src/main/java/bl4ckscor3/mod/globalxp/openmods/utils/EnchantmentUtils.java:34]: From a quick glance, I think XP loss can occur here when the player is removing an amount of XP from the block that would result in them having more XP than the integer limit. I have to properly test this when I have more time to look at my mods.

Comment (minecraft552) [reply src/main/java/bl4ckscor3/mod/globalxp/openmods/utils/EnchantmentUtils.java:34]: ~~This should be correct, however the amount of XP levels for that to happen is 238,609,312 according to the minecraft wiki https://minecraft.wiki/w/Experience#Useful_numbers, so I ignored it on purpose~~
edit: NVM, I miscalculated, let me redo calculations

Comment (bl4ckscor3) [reply src/main/java/bl4ckscor3/mod/globalxp/openmods/utils/EnchantmentUtils.java:34]: If you have more than the integer limit of XP on your player, Minecraft works just fine even though this value gets stuck at the max integer value. (Testable by setting your XP to 0 points and levels, then running `/xp add @s 2147483647` twice). So maybe this is a non-issue, but still worth looking into.

Comment (minecraft552) [reply src/main/java/bl4ckscor3/mod/globalxp/openmods/utils/EnchantmentUtils.java:34]: It does break at some point after 47721862 levels (which is not anywhere near the vanilla maximum levels), so I am going to debug it.

**Comment by bl4ckscor3** [comment]
I'm not even 100% sure I want this change. Maybe it's better to fix the overflow instead of increasing the limit, the latter of which would just move that bug to an, albeit ridiculously much, higher value anyway. One argument for not increasing the block's limit is that vanilla also limits XP on the player to the integer limit (see for example the experience command).

**Comment by minecraft552** [comment]
well, that's XP, but not XP levels, as you know minecraft nowadays stores your XP in levels and progress, so the level limit is higher (technically int limit in code, but as stated only goes up to ~238 million in vanilla survival). Either way, it's your mod so it's up to you, I don't think it hurts to add it, but on the other hand this feature will likely be used by only 1 or 2 people, so there's not much reason to add it.

**Comment by minecraft552** [comment]
I can also make a PR for the overflow if you want btw

**Comment by bl4ckscor3** [comment]
[MC-3266 ("Your experience level overflows once it exceeds the integer limit")](https://bugs.mojang.com/browse/MC-3266) just got marked as fixed, let's see in the next snapshot how they addressed this and I'll have another look here.

**Comment by minecraft552** [comment]
Well that's related to XP levels and not XP points which is how things are stored for XPBlocks, so it shouldn't be very relevant.
speaking of which, converting the storage to XP levels and points separately should also fix all our issues up to that 238 million limit

**Comment by bl4ckscor3** [comment]
> Well that's related to XP levels and not XP points

I'm fully aware of that. That doesn't change the fact that there may be relevant changes made that would influence my decision on this.

**Comment by bl4ckscor3** [comment]
I decided to cap the maximum XP at the integer limit.

**Comment by bl4ckscor3** [comment]
See 1e57b67c719813f69b02db6473394bb98b7fa769

**Pull request closed without merging**